diffoscope: move unfree dependencies behind a enableUnfree flag

RaitoBezarius opened the pull request:
###### Description of changes

`diffoscope` no longer relies on unfree dependencies.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

RaitoBezarius committed:
diffoscope: move unfree dependencies behind a enableUnfree flag

SuperSandro2000 commented on pkgs/tools/misc/diffoscope/default.nix:
```diff
@@ -3,7 +3,8 @@
 , gzip, html2text, hdf5, imagemagick, jdk, libarchive, libcaca, llvm, lz4, mono, ocaml, oggvideotools, openssh, openssl, pdftk, pgpdump, poppler_utils, procyon, qemu, R
 , radare2, sng, sqlite, squashfsTools, tcpdump, ubootTools, odt2txt, unzip, wabt, xmlbeans, xxd, xz, zip, zstd
-, enableBloat ? false
+, enableBloat ? !stdenv.isDarwin
```
What is the reasoning behind this change?

SuperSandro2000 commented on pkgs/tools/misc/diffoscope/default.nix:
```diff
@@ -3,7 +3,8 @@
 , gzip, html2text, hdf5, imagemagick, jdk, libarchive, libcaca, llvm, lz4, mono, ocaml, oggvideotools, openssh, openssl, pdftk, pgpdump, poppler_utils, procyon, qemu, R
 , radare2, sng, sqlite, squashfsTools, tcpdump, ubootTools, odt2txt, unzip, wabt, xmlbeans, xxd, xz, zip, zstd
-, enableBloat ? false
+, enableBloat ? !stdenv.isDarwin
```
```suggestion
, enableBloat ? true
```
if minimal sets this to false, it is less confusing if we default this to true.

Also I don't think it makes sense to always get the diffoscopeMinimal on darwin and on linux you get the full one with diffoscope. That's almost always unexpected. If packages don't work on darwin then they should be rather moved to linux only like it is already done for some.

RaitoBezarius replied:
cc @alyssais who suggested to me this change.

alyssais replied:
> Also I don't think it makes sense to always get the diffoscopeMinimal on darwin and on linux you get the full one with diffoscope. That's almost always unexpected. If packages don't work on darwin then they should be rather moved to linux only like it is already done for some.

This makes sense to me.

RaitoBezarius replied:
Let's see what breaks on Darwin and let's move it then

SuperSandro2000 commented on pkgs/tools/misc/diffoscope/default.nix:
```diff
@@ -54,8 +55,9 @@ python3Packages.buildPythonApplication rec {
       python-magic progressbar33 pypdf2 tlsh
     ])
     ++ lib.optionals stdenv.isLinux [ python3Packages.pyxattr python3Packages.rpm acl cdrkit dtc ]
+    ++ lib.optionals (enableBloat && enableUnfree) [ apktool ]
```
```suggestion
```

alyssais replied:
I think it makes more sense to make it require both options.  It's not something most users are likely to need, so it's bloat, and it's unfree.  Just because we don't currently have any non-bloat unfree doesn't mean that that category doesn't make sense.

SuperSandro2000 replied:
I just moved it within the enableBloat conditional path

RaitoBezarius replied:
Done

SuperSandro2000 commented on pkgs/tools/misc/diffoscope/default.nix:
```diff
@@ -54,8 +55,9 @@ python3Packages.buildPythonApplication rec {
-      abootimg apksigcopier apksigner apktool cbfstool colord enjarify ffmpeg fpc ghc ghostscriptX giflib gnupg gnumeric
+      abootimg apksigcopier apksigner cbfstool colord enjarify ffmpeg fpc ghc ghostscriptX giflib gnupg gnumeric
       hdf5 imagemagick libcaca llvm jdk mono ocaml odt2txt oggvideotools openssh pdftk poppler_utils procyon qemu R tcpdump ubootTools wabt radare2 xmlbeans
     ] ++ (with python3Packages; [ androguard binwalk guestfs h5py pdfminer-six ]));
```
```suggestion
    ] ++ (with python3Packages; [ androguard binwalk guestfs h5py pdfminer-six ])
    ++ lib.optionals enableUnfree [ apktool ]);
```

RaitoBezarius replied:
Done, let me know if I have done it wrong

davidak commented on pkgs/tools/misc/diffoscope/default.nix:
```diff
@@ -3,7 +3,8 @@
 , radare2, sng, sqlite, squashfsTools, tcpdump, ubootTools, odt2txt, unzip, wabt, xmlbeans, xxd, xz, zip, zstd
-, enableBloat ? false
+, enableBloat ? !stdenv.isDarwin
+, enableUnfree ? false
```
there probably is no policy, but shouldn't this default to `nixpkgs.config.allowUnfree`?

and in a good UX, shouldn't it also work when allowing the specific dependency with `allowUnfreePredicate` or `allowlistedLicenses` etc?

maybe this needs more design and a proper implementation and policy for nixpkgs

RaitoBezarius replied:
it would create the surprising behavior where depending on allowUnfree or not, you would have to compile or have a different package from the named one

davidak replied:
>you would have to compile

yes, i thought about that

but documentation already sais:

>Note that we are not able to test or build unfree software on Hydra due to policy.

source: https://nixos.org/manual/nixpkgs/stable/#chap-packageconfig

so shouldn't surprise

>have a different package from the named one

you mean when one person with `nixpkgs.config.allowUnfree = true;` installs diffoscope, they have a different package than one person with `false`?

it's kind of implicit, but i can see that as expected behavior

it would also be different when you override it explicitly

i think this is ok for now, but it's an interesting topic for the Nixpkgs Architecture Team to create a proper interface for packages

RaitoBezarius replied:
> it's kind of implicit, but i can see that as expected behavior

Beyond "expected" or not, it makes it impossible to **not rebuild** diffoscope if you have allowUnfree without overriding the package, which is, IMHO, very bad UX.

A user probably want the cache, otherwise you will plenty of people saying "why am I compiling this" (unfree=true) while people will say "huh no it works with me" (unfree=false).

So it is definitely surprising to my eyes.

alyssais replied:
> there probably is no policy, but shouldn't this default to `nixpkgs.config.allowUnfree`?

As far as I can tell, we 100% consistently do not do this in Nixpkgs.

davidak commented on pkgs/tools/misc/diffoscope/default.nix:
```diff
@@ -54,8 +55,9 @@ python3Packages.buildPythonApplication rec {
       python-magic progressbar33 pypdf2 tlsh
     ])
     ++ lib.optionals stdenv.isLinux [ python3Packages.pyxattr python3Packages.rpm acl cdrkit dtc ]
+    ++ lib.optionals (enableBloat && enableUnfree) [ apktool ]
```
does this imply apktool is unfree? according to [the package search](https://search.nixos.org/packages?channel=unstable&from=0&size=50&sort=relevance&type=packages&query=apktool) it has [Apache License 2.0](https://spdx.org/licenses/Apache-2.0.html)

RaitoBezarius replied:
it pulls an unfree dependency I suppose related to Android

davidak replied:
Can you add as a comment to the code which one, so it can be understood later.

RaitoBezarius replied:
got it

SuperSandro2000 replied:
build-tools is unfree which comes from ``inherit (androidenv.androidPkgs_9_0) build-tools;`` and for that you need to accept the androidsdk license with an extra attribute within config.

raboof reviewed (commented):
LGTM, with the difference that `enableBloat` should just default to `true` (and perhaps explicitly mark the bloat+darwin combination as broken/unsupported, assuming that's still the case)

RaitoBezarius commented:
Result of `nixpkgs-review pr 221861` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>diffoscope</li>
    <li>diffoscope.dist</li>
    <li>diffoscope.man</li>
  </ul>
</details>

RaitoBezarius commented:
Result of `nixpkgs-review pr 221861` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)

RaitoBezarius commented:
Result of `nixpkgs-review pr 221861` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)

RaitoBezarius committed:
diffoscope: fix build on Darwin

Moved packages requiring x86_64-linux, x86_64-darwin into their proper arrays.

RaitoBezarius commented:
Result of `nixpkgs-review pr 221861` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages failed to build:</summary>
  <ul>
    <li>diffoscope</li>
    <li>diffoscope.dist</li>
    <li>diffoscope.man</li>
  </ul>
</details>

RaitoBezarius commented:
`┃    ├─ ⚠︎ oggvideotools-0.9.1-aarch64-darwin on [2] failed with exit code 1 after ⏱︎ 6s` is broken on aarch64-darwin.

RaitoBezarius commented:
Result of `nixpkgs-review pr 221861` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages failed to build:</summary>
  <ul>
    <li>diffoscope</li>
    <li>diffoscope.dist</li>
    <li>diffoscope.man</li>
  </ul>
</details>

RaitoBezarius commented:
Same for x86_64-darwin.

RaitoBezarius commented:
Marking them as broken.

RaitoBezarius committed:
oggvideotools: mark it as broken on Darwin

RaitoBezarius commented:
cc @SuperSandro2000 as you are maintainer of the package

alyssais commented on pkgs/tools/misc/diffoscope/default.nix:
```diff
@@ -54,10 +55,20 @@ python3Packages.buildPythonApplication rec {
+      # oggvideotools is broken on Darwin, please put it back when it will be fixed?
+      ++ lib.optionals stdenv.isLinux [ abootimg cbfstool colord ubootTools python3Packages.guestfs oggvideotools ]
+      # This doesn't work on aarch64-darwin
+      ++ lib.optionals (stdenv.isx86_64 || stdenv.isLinux) [ gnumeric ]
```
Unless it does something specifically to only work on aarch64 on Linux, this doesn't mean it should be disabled for all non-Linux OSes on aarch64:

```suggestion
      ++ lib.optionals (stdenv.hostPlatform.system != "aarch64-darwin") [ gnumeric ]
```

RaitoBezarius replied:
Right

RaitoBezarius replied:
Done

alyssais commented on pkgs/tools/misc/diffoscope/default.nix:
```diff
@@ -54,10 +55,20 @@ python3Packages.buildPythonApplication rec {
+      ]
+      ++ (with python3Packages; [ androguard binwalk h5py pdfminer-six ])
+      # oggvideotools is broken on Darwin, please put it back when it will be fixed?
+      ++ lib.optionals stdenv.isLinux [ abootimg cbfstool colord ubootTools python3Packages.guestfs oggvideotools ]
```
It would be nice to filter based on `lib.meta.availableOn`, to make this more accurate / maintainable.

(But please don't filter on meta.broken, as that's more likely to flip back and forth over time and surprise people.)

RaitoBezarius replied:
Would you advise to filter the whole list or filter only parts of the list? Would you make this part of this PR or can we revisit this in another PR?

alyssais replied:
I think it should be part of this PR, because you're introducing the `isLinux` condition in this PR, right?

RaitoBezarius replied:
Correct

SuperSandro2000 replied:
> It would be nice to filter based on `lib.meta.availableOn`, to make this more accurate / maintainable.

That sounds like a really good idea to me.

SuperSandro2000 replied:
Can we also filter unfree packages like that?

RaitoBezarius replied:
> Can we also filter unfree packages like that?

I don't think it's a good idea to do it, see one of the resolved discussion on that with @davidak

RaitoBezarius replied:
I implemented the proposal, let me know if you see any other way to do it.

SuperSandro2000 replied:
> I don't think it's a good idea to do it, see one of the resolved discussion on that with @davidak

We don't need to make it dependend on config.allowUnfree

SuperSandro2000 commented on pkgs/tools/misc/oggvideotools/default.nix:
```diff
@@ -43,5 +43,6 @@ stdenv.mkDerivation rec {
     homepage = "http://www.streamnik.de/oggvideotools.html";
     license = licenses.gpl2Only;
     maintainers = with maintainers; [ SuperSandro2000 ];
+    broken = stdenv.isDarwin;
```
Can we add a very short comment with the reason?

RaitoBezarius replied:
This is done

RaitoBezarius committed:
python3Packages.guestfs: platforms are the ones libguestfs supports

This ensures the proper propagation for `lib.meta.availableOn` to work fine.

RaitoBezarius committed:
python3Packages.pyxattr: platforms are the ones xattr supports

This ensures the proper propagation for `lib.meta.availableOn` to work.

RaitoBezarius committed:
diffoscope: introduce `lib.meta.availableOn stdenv.hostPlatform` for "plugins"

This makes it easier to add new plugins without having to worry whether they are supported on Darwin, aarch64-*, etc.

As long as they are properly tagged through their `platforms` meta attribute (or `badPlatforms`).

Broken packages needs to be separated to avoid silent breakages which we would not notice.

RaitoBezarius commented:
Result of `nixpkgs-review pr 221861` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>6 packages built:</summary>
  <ul>
    <li>diffoscope</li>
    <li>diffoscope.dist</li>
    <li>diffoscope.man</li>
    <li>diffoscopeMinimal</li>
    <li>diffoscopeMinimal.dist</li>
    <li>diffoscopeMinimal.man</li>
  </ul>
</details>
